src/lib/dnssd/tests/TestIncrementalResolve.cpp

```diff
@@ -0,0 +1,427 @@
+
+    NL_TEST_ASSERT(inSuite, record.Parse(srv_data, packet));
+}
+
+/// Convenience method to have a  serialized QName:
+///
+/// static const uint8_t kData[] = "datahere\00";
+///  AsSerializedQName(kData);
+///
+/// NOTE: this MUST be using the string "" format to add an extra NULL
+/// terminator that this method discards.
+template <size_t N>
+static SerializedQNameIterator AsSerializedQName(const uint8_t (&v)[N])
+{
+    // NOTE: the -1 is because we format these items as STRINGS and that
+    // appends an extra NULL terminator
+    return SerializedQNameIterator(BytesRange(v, v + N - 1), v);
+}
+
+void CallOnRecord(nlTestSuite * inSuite, IncrementalResolver & resolver, const ResourceRecord & record)
+{
+    uint8_t headerBuffer[HeaderRef::kSizeBytes] = {};
+    HeaderRef dummyHeader(headerBuffer);
+
+    uint8_t dataBuffer[256];
+    chip::Encoding::BigEndian::BufferWriter output(dataBuffer, sizeof(dataBuffer));
+    RecordWriter writer(&output);
+
+    NL_TEST_ASSERT(inSuite, record.Append(dummyHeader, ResourceType::kAnswer, writer));
+    NL_TEST_ASSERT(inSuite, writer.Fit());
+
+    ResourceData resource;
+    BytesRange packet(dataBuffer, dataBuffer + sizeof(dataBuffer));
+    const uint8_t * _ptr = dataBuffer;
+    NL_TEST_ASSERT(inSuite, resource.Parse(packet, &_ptr));
+    NL_TEST_ASSERT(inSuite, resolver.OnRecord(resource, packet) == CHIP_NO_ERROR);
+}
+
+void TestStoredServerName(nlTestSuite * inSuite, void * inContext)
+{
+
+    StoredServerName name;
+
+    // name should start of as cleared
+    NL_TEST_ASSERT(inSuite, !name.Get().Next());
+
+    // Data should be storable in server name
+    NL_TEST_ASSERT(inSuite, name.Set(AsSerializedQName(kTestOperationalName)) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, name.Get() == AsSerializedQName(kTestOperationalName));
+    NL_TEST_ASSERT(inSuite, name.Get() != AsSerializedQName(kTestCommissionerNode));
+    NL_TEST_ASSERT(inSuite, name.Get() != AsSerializedQName(kTestCommissionableNode));
+
+    NL_TEST_ASSERT(inSuite, name.Set(AsSerializedQName(kTestCommissionerNode)) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, name.Get() != AsSerializedQName(kTestOperationalName));
+    NL_TEST_ASSERT(inSuite, name.Get() == AsSerializedQName(kTestCommissionerNode));
+    NL_TEST_ASSERT(inSuite, name.Get() != AsSerializedQName(kTestCommissionableNode));
+
+    NL_TEST_ASSERT(inSuite, name.Set(AsSerializedQName(kTestCommissionableNode)) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, name.Get() != AsSerializedQName(kTestOperationalName));
+    NL_TEST_ASSERT(inSuite, name.Get() != AsSerializedQName(kTestCommissionerNode));
+    NL_TEST_ASSERT(inSuite, name.Get() == AsSerializedQName(kTestCommissionableNode));
+
+    {
+        // setting to a too long value should reset it
+        uint8_t largeBuffer[256];
+        for (unsigned i = 0; i < 20; i++)
+        {
+            memcpy(largeBuffer + i * 8, "\07abcd123", 8);
+            largeBuffer[(i + 1) * 8] = 0;
```

Comment:
Just use a `chip::Encoding::BigEndian::BufferWrite` and call Put on it N times to avoid all this arithmetic